spark/src/main/scala/geotrellis/spark/io/cog/COGLayerReader.scala

```diff
@@ -54,16 +65,184 @@ abstract class COGLayerReader[ID] extends Serializable {
+    V <: CellGrid: GeoTiffReader: ClassTag
+  ](layerId: ID): BoundLayerQuery[K, TileLayerMetadata[K], RDD[(K, V)] with Metadata[TileLayerMetadata[K]]] =
+    new BoundLayerQuery(new LayerQuery, read(layerId, _))
+
+  def query[
+    K: SpatialComponent: Boundable: JsonFormat: ClassTag,
+    V <: CellGrid: GeoTiffReader: ClassTag
+  ](layerId: ID, numPartitions: Int): BoundLayerQuery[K, TileLayerMetadata[K], RDD[(K, V)] with Metadata[TileLayerMetadata[K]]] =
+    new BoundLayerQuery(new LayerQuery, read(layerId, _, numPartitions))
+
+  def querySubsetBands[
+    K: SpatialComponent: Boundable: JsonFormat: ClassTag
+  ](layerId: ID, targetBands: Seq[Int]): BoundLayerQuery[K, TileLayerMetadata[K], RDD[(K, Array[Option[Tile]])] with Metadata[TileLayerMetadata[K]]] =
+    new BoundLayerQuery(new LayerQuery, readSubsetBands(layerId, targetBands, _))
+
+  def querySubsetBands[
+    K: SpatialComponent: Boundable: JsonFormat: ClassTag
+  ](layerId: ID, targetBands: Seq[Int], numPartitions: Int): BoundLayerQuery[K, TileLayerMetadata[K], RDD[(K, Array[Option[Tile]])] with Metadata[TileLayerMetadata[K]]] =
+    new BoundLayerQuery(new LayerQuery, readSubsetBands(layerId, targetBands, _, numPartitions))
+
+  private def crop[V <: CellGrid: GeoTiffReader: ClassTag](
+    geoTiff: GeoTiff[V],
+    gridBounds: Seq[GridBounds]
+  ): Iterator[(GridBounds, V)] =
+    geoTiff.crop(gridBounds)
+
+  private def produceCropBands(
+    targetBands: Seq[Int]
+  ): (GeoTiff[MultibandTile], Seq[GridBounds]) => Iterator[(GridBounds, Array[Option[Tile]])] =
+    (geoTiff: GeoTiff[MultibandTile], gridBounds: Seq[GridBounds]) => cropBands(geoTiff, gridBounds, targetBands)
+
+  private def cropBands(
+    geoTiff: GeoTiff[MultibandTile],
+    gridBounds: Seq[GridBounds],
+    bands: Seq[Int]
+  ): Iterator[(GridBounds, Array[Option[Tile]])] = {
+    // We first must determine which bands are valid and which are not
+    // before doing the crop in order to avoid band subsetting errors
+    // and/or loading unneeded data.
+    val targetBandsWithIndex: Array[(Int, Int)] =
+      bands
+        .zipWithIndex
+        .filter { case (band, _) =>
+          band >= 0 && band < geoTiff.bandCount
+        }
+        .toArray
+
+    val (targetBands, targetBandsIndexes) = targetBandsWithIndex.unzip
+
+    val croppedTilesWithGridBounds: Iterator[(GridBounds, Array[Tile])] =
+      geoTiff
+        .tile
+        .cropBands(gridBounds, targetBands)
+        .map { case (k, v) => k -> v.bands.toArray }
+
+    val croppedTilesWithBandIndexes: Iterator[(GridBounds, Array[(Int, Tile)])] =
+      croppedTilesWithGridBounds
+        .map { case (k, v) => k -> targetBandsIndexes.zip(v) }
+
+    croppedTilesWithBandIndexes.map { case (k, v) =>
+      val returnedBands = Array.fill[Option[Tile]](bands.size)(None)
+      for ((index, band) <- v) {
+        returnedBands(index) = Some(band)
+      }
+
+      k -> returnedBands
+    }
+  }
+
+  private def transformKeyBounds[
```

Comment:
Yeah, I was thinking they were a bit out of place there. I don't think that `produceCropBands`, `cropBands`, and `crop` really belong in the raster classes. Since those a pretty specific to COGs. What if we made a `COGTransformations` object that'll have these methods?

Yeah, I thought `transformKeyBounds` elsewhere. I'll look to see where it is.